### PR TITLE
Improve Move Objects Update Feature

### LINKED ISSUE #464: Improve Feature for Moving Shells
#### Description:
Currently, moving shells (Walls or Slabs) in ETABS via the BHoM is not carried out correctly.
	- BHoM Error Message: "The Etabs API does not allow for updating of the geometry of panels." 
	  "This includes the external edges as well as the openings."
	  "To update the panel geometry, delete the existing panel you want to update and create a new one."
It looks that the ETABS API provides a cEditGeneral.Move Method that moves any PointObj/FrameObj/AreaObj(?) that is selected (obj.SetSelected(....)). 
        - Test the method and see if possible to develop an algorithm in the BHoM based on it, thus, allowing to sort out the issue above.

Move Shells Issue

### PR DESCRIPTION
**Pull request by GCRA101**
### Issues addressed by this PR

 Closes #464 

GH Script and Model Preview

VS - Node Move Algorithm After

ETABS Toolkit now able to perform translational movement of ETABS models upon Update Request much faster than before (up to 7 times faster as per Test ETABS model below - i.e.: from 7 minutes to 1 minute max).
Whenever rotations or distorsions of ETABS objects are concerned, unfortunately ETABS API limitations still prevent BHoM from carrying out a correct update of ETABS panel objects.
More work is required to find out a solution to this problem.

 ### Test files
Video Demonstration - Issue
https://burohappold.sharepoint.com/:v:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23476-ImproveMoveObjectsAlgorithm/Moving%20BHoM%20Algorithm%20Issue.mp4?csf=1&web=1&e=XrufhL
Video Demonstration - Solution
https://burohappold.sharepoint.com/:v:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23476-ImproveMoveObjectsAlgorithm/Moving%20BHoM%20Algorithm%20Sorted.mp4?csf=1&web=1&e=ZlxeGo
Grasshopper File
https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23476-ImproveMoveObjectsAlgorithm/[TestScript.gh](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23476-ImproveMoveObjectsAlgorithm/TestScript.gh?csf=1&web=1&e=1smiHN)?csf=1&web=1&e=1smiHN
ETABS File
https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23476-ImproveMoveObjectsAlgorithm/Test%20ETABS%20Model.EDB?csf=1&web=1&e=X61Bhi

 ### Changelog

- Use of Hash Tables to store and access faster data for moving nodes
- Use of Streams on Hash Tables to run the ETABS API .move() method faster on nodes grouped together

### CODE REVIEW

**Review by Chrisshort92 — DISMISSED** [review]
Functionality tested with provided test script and file. Files working as expected therefore PR approved for merge.

**Comment by Chrisshort92** [comment]
@BHoMBot check required

**Comment by peterjamesnugent** [comment]
@BHoMBot check copyright-compliance

File: Etabs_Adapter/CRUD/Update/Node.cs [hunk 1]
```diff
@@ -27,6 +27,9 @@
 using BH.oM.Structure.Elements;
 using BH.oM.Structure.Constraints;
 using BH.Engine.Adapters.ETABS;
+using System.Collections;
+using System.Xml.Linq;
```
Comment (peterjamesnugent) [inline Etabs_Adapter/CRUD/Update/Node.cs:31]: Is this needed?

**Review by peterjamesnugent — CHANGES_REQUESTED** [review]
Few minor changes.

File: Etabs_Adapter/CRUD/Update/Node.cs [hunk 2]
```diff
@@ -44,44 +47,97 @@ public partial class ETABSAdapter : BHoMAdapter
...
-            double factor = DatabaseLengthUnitFactor();
+            double factor = DatabaseLengthUnitFactor();                                        // θ(1)
 
-            Engine.Structure.NodeDistanceComparer comparer = AdapterComparers[typeof(Node)] as Engine.Structure.NodeDistanceComparer;
+            Engine.Structure.NodeDistanceComparer comparer = AdapterComparers[typeof(Node)]    // θ(1)
+                as Engine.Structure.NodeDistanceComparer;
+
+            Dictionary<double, List<string>> Δx = new Dictionary<double, List<string>>();      // θ(1)
+            Dictionary<double, List<string>> Δy = new Dictionary<double, List<string>>();      // θ(1)
+            Dictionary<double, List<string>> Δz = new Dictionary<double, List<string>>();      // θ(1)
```
[the remark this answers is not in the document]
Comment (GCRA101) [reply Etabs_Adapter/CRUD/Update/Node.cs:60]: @peterjamesnugent, thanks for the tip.
delta replaced with d letter in latest commit of the code

File: Etabs_Adapter/CRUD/Update/Node.cs (hunk 1 above)
Comment (GCRA101) [reply Etabs_Adapter/CRUD/Update/Node.cs:31]: @peterjamesnugent, thanks for the comment.
Libraries not used. References deleted from the code in latest commit

File: Etabs_Adapter/ETABSAdapter.cs [hunk 3]
```diff
@@ -158,11 +158,14 @@ private bool ForceRefresh()
...
-            m_model.EditGeneral.Move(-1, 0, 0);
-            m_model.SelectObj.ClearSelection();
+            //m_model.SelectObj.All();
+            //m_model.EditGeneral.Move(1, 0, 0);
+            //m_model.SelectObj.All();
+            //m_model.EditGeneral.Move(-1, 0, 0);
+            //m_model.SelectObj.ClearSelection();
+
+            m_model.View.RefreshView();
+            m_model.View.RefreshWindow();
```
[the remark this answers is not in the document]
Comment (GCRA101) [reply Etabs_Adapter/ETABSAdapter.cs:168]: @peterjamesnugent, comments removed as suggested

File: Etabs_Adapter/CRUD/Update/Node.cs [hunk 4]
```diff
@@ -44,44 +47,97 @@ public partial class ETABSAdapter : BHoMAdapter
...
+                // 2. Move all selected nodes by same ΔZ
+                m_model.EditGeneral.Move(0, 0, (double)kvp.Key);
+                // 3. Deselect all selected nodes
+                kvp.Value.ForEach(pplbl => m_model.PointObj.SetSelected(pplbl.ToString(), false));
+            });
+
             return success;
         }
 
+        /* Computational Cost: T(n)= θ(n) */
```
[the remark this answers is not in the document]
Comment (GCRA101) [reply Etabs_Adapter/CRUD/Update/Node.cs:139]: @peterjamesnugent, comment line removed as suggested

**Review by peterjamesnugent — APPROVED** [review]
Happy from a code side, @Chrisshort92 can you do a functionality review please.

**Comment by peterjamesnugent** [comment]
@BHoMBot check required
@BHoMBot check copyright-compliance

**Comment by peterjamesnugent** [comment]
@BHoMBot check installer

**Pull request merged**